numpy/core/src/umath/ufunc_object.c

```diff
@@ -781,9 +787,11 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
 #if !defined(NPY_PY3K)
             if (DEPRECATE("unorderable dtypes; returning scalar but in "
                           "the future this will be an error") < 0) {
-                return -1;
+                goto fail;
             }
 #endif
+            Py_DECREF(out_op[0]);
+            Py_DECREF(out_op[1]);
```

Comment:
Should this do the full cleanup? Could add a `fail_pre_kwargs` label, if we want to avoid unnecessary cleanup.

Of course, really the huge hack section of this function should be extracted from the rest of it.

Reply:
Initially I had a `goto cleanup_out_op` and passed on a `retval`, but then I found that it just became confusing, and might hinder ripping this whole part out (which, as you said, should be done, but better not in this PR...). Fortunately, here we know exactly what has been done up to this point (indeed, this is partially why I moved the `nin == 2` to the if-statement; I wanted to be absolutely sure I decref'd just the right number of operands, even if someone gave their ufunc a name that matches but has more than 2 inputs).

Reply:
More for completeness: See #11282